src/backend/distributed/worker/worker_create_or_replace.c

```diff
@@ -0,0 +1,179 @@
+#include "fmgr.h"
+#include "nodes/makefuncs.h"
+#include "nodes/nodes.h"
+#include "parser/parse_type.h"
+#include "tcop/dest.h"
+#include "tcop/utility.h"
+#include "utils/builtins.h"
+#include "utils/regproc.h"
+
+#include "distributed/commands.h"
+#include "distributed/commands/utility_hook.h"
+#include "distributed/deparser.h"
+#include "distributed/metadata/distobject.h"
+#include "distributed/worker_protocol.h"
+
+static Node * CreateStmtByObjectAddress(const ObjectAddress *address);
+static DropStmt * drop_stmt_from_object_create(Node *createStmt);
+
+
+PG_FUNCTION_INFO_V1(worker_create_or_replace);
+
+
+/*
+ * worker_create_or_replace(statement text)
+ *
+ * function is called, by the coordinator, with a CREATE statement for an object. This
+ * function implements the CREATE ... IF NOT EXISTS functionality for objects that do not
+ * have this functionality or where their implementation is not sufficient.
+ *
+ * Besides checking if an object of said name exists it tries to compare the object to be
+ * created with the one in the local catalog. If there is a difference the on in the local
+ * catalog will be renamed after which the statement can be executed on this worker to
+ * create the object.
+ *
+ * Renaming has two purposes
+ *  - free the identifier for creation
+ *  - non destructive if there is data store that would be destroyed if the object was
+ *    used in a table on this node, eg. types. If the type would be dropped with a cascade
+ *    it would drop any column holding user data for this type.
+ */
+Datum
+worker_create_or_replace(PG_FUNCTION_ARGS)
+{
+	text *sqlStatementText = PG_GETARG_TEXT_P(0);
+	const char *sqlStatement = text_to_cstring(sqlStatementText);
+	const ObjectAddress *address = NULL;
+	Node *parseTree = ParseTreeNode(sqlStatement);
+
+	/*
+	 * since going to the drop statement might require some resolving we will do a check
+	 * if the type actually exists instead of adding the IF EXISTS keyword to the
+	 * statement.
+	 */
+	address = GetObjectAddressFromParseTree(parseTree, true);
+	if (ObjectExists(address))
+	{
+		Node *localCreateStmt = NULL;
+		const char *localSqlStatement = NULL;
+		DropStmt *dropStmtParseTree = NULL;
+
+		localCreateStmt = CreateStmtByObjectAddress(address);
+		localSqlStatement = DeparseTreeNode(localCreateStmt);
+		if (strcmp(sqlStatement, localSqlStatement) == 0)
+		{
+			/*
+			 * TODO string compare is a poor mans comparison, but calling equal on the
+			 * parsetree's returns false because there is extra information list character
+			 * position of some sort
+			 */
+
+			/*
+			 * parseTree sent by the coordinator is the same as we would create for our
+			 * object, therefore we can omit the create statement locally and not create
+			 * the object as it already exists.
+			 *
+			 * We let the coordinator know we didn't create the object.
+			 */
+			PG_RETURN_BOOL(false);
+		}
+
+		/* TODO don't drop, instead rename as described in documentation */
```

Comment:
As far as I remember from in person discussions, you're working on this, right?

Reply:
Let's punt on that until after merge to unblock others.

Reply:
I have the test written, implementation almost, I can leave it out for now but is essential to add before 9.0

Reply:
issue for tracking https://github.com/citusdata/citus/issues/2953